### ISSUE TITLE
Cache stampede prevention for empty cache

### ISSUE THREAD

**Issue by sobhan-m94** [yiisoft/cache#123]
The current implementation only works when the data is already cached. In this case, every time the data is read from the cache, the algorithm calculates the expiration time and, if key is expired, recompute the cache. But imagine the cache is empty(for the first time) in high traffic. If we get a thousand simultaneous requests, they all hit the empty cache and the system calculates the data for all of them. Which again causes the cache stampede problem.
Is there a way to prevent the calculation of it for all requests in the first time, when our key and data are not cached? 
For example lock the key or anything else ?

```php
$now = microtime(true);
$delta = ceil(1000 * ($now - $this->updated)) / 1000;
$expired = $now - $delta * $beta * log(random_int(1, PHP_INT_MAX) / PHP_INT_MAX);

return $this->expiry <= $expired;
```

![ezgif-4-17a8ea5945](https://github.com/yiisoft/cache/assets/44964722/a56804ca-ee22-4e37-b440-97172db99b10)

**Comment by samdark** [comment]
One of the obvious solutions — warmup. How would you solve it?

**Comment by samdark** [comment]
The situation you have isn't "Cache stampede" but "[Thundering herd](https://en.wikipedia.org/wiki/Thundering_herd_problem)".

**Comment by sobhan-m94** [comment]
Cache stampede, also known as a thundering herd : 

When a large number of concurrent requests attempt to access the same data from the cache simultaneously, but find it **missing or expired.**

https://www.thetechplatform.com/amp/how-to-prevent-cache-stampede-thundering-herd-problems

**Comment by samdark** [comment]
Yeah. Seems to be different names used for the same thing. I found that "thundering herd" if usually used for empty cache situation and stampede is used for when many cache items are expiring at the same time.

**Comment by sobhan-m94** [comment]
Why locking is not good idea ?

**Comment by samdark** [comment]
1. Higher latency.
2. Reduced concurrency.
3. Potential for deadlocks.

**Comment by samdark** [comment]
I see no other immediate solutions to empty-cache problem i.e. around https://github.com/yiisoft/cache/blob/master/src/Cache.php#L79 we can do the following:

```php
public function getOrSet(
        mixed $key,
        callable $callable,
        DateInterval|int|null $ttl = null,
        Dependency $dependency = null,
        float $beta = 1.0
    ) {
        $key = $this->keyNormalizer->normalize($key);
        /** @var mixed */
        $value = $this->getValue($key, $beta);

       if ($value) {
           return $value;
       }

       // We'll get additional N cache reads. Better than N reads from DB but...
       while ($lease = $this->psr->getRaw('lease_' . $key)) {
          usleep(500); // The higher the value the worse latency is.
       }

       // How to choose correct expiration time here?
       // That will double the amount of keys and reads/writes for cache overall.
       $this->psr->set('lease_' . $key, 1, max($ttl, 1500));
       return $this->setAndGet($key, $callable, $ttl, $dependency);
    }
```

**Comment by vjik** [comment]
It's should be optional and disabled for cases where user own resolve this problem on application level.
